Public/Src/Pips/Dll/Operations/Process.cs

```diff
@@ -765,6 +793,119 @@ public bool TryComputePipUniqueOutputHash(PathTable pathTable, out long pipUniqu
+
+            foreach (var file in SourceChangeAffectedOutputFiles)
+            {
+                foreach (var directory in DirectoryDependencies)
+                {
+                    if (directory.Path.TryGetRelative(pathTable, file, out var relativePath))
+                    {
+                        changeAffectedInputs.Add(file.GetName(pathTable).ToString(pathTable.StringTable));
+                    }
+                }
+            }
+
+            foreach (var outDir in SourceChangeAffectedOutputDirectories)
+            {
+                foreach (var inDir in DirectoryDependencies)
+                {
+                    if (outDir.TryGetRelative(pathTable, inDir.Path, out var relativePath))
+                    {
+                        FileUtilities.EnumerateDirectoryEntries(
+                            inDir.Path.ToString(pathTable), 
+                            true, 
+                            (dir, fileName, attributes) =>
+                            {
+                                if (attributes == FileAttributes.Archive)
+                                {
+                                    changeAffectedInputs.Add(fileName);
+                                }
+                            });
+                    }
+                }
+            }
+
+            return ReadOnlyArray<string>.FromWithoutCopy(changeAffectedInputs.ToArray());
+        }
+
+        /// <inheritdoc />
+        public override bool IsOutputAffectedBySourceChange(
+            ReadOnlyArray<AbsolutePath> dynamicallyObservedFiles,
+            ReadOnlyArray<AbsolutePath> dynamicallyObservedEnumerations,
+            PathTable pathTable,
+            IReadOnlyCollection<AbsolutePath> sourceChangeAffectedOutputFiles = null,
+            IReadOnlyCollection<AbsolutePath> sourceChangeAffectedOutputDirectroies = null)
+        {
+            // sourceChangeAffectedOutputFiles and sourceChangeAffectedOutputDirectroies 
+            // are initilized with the source change list from GBR.
+            // If it is null, that means no file or directory change for this build.
+            // Thus no output is affected by the change.
+            if (!sourceChangeAffectedOutputFiles.Any() && !sourceChangeAffectedOutputDirectroies.Any())
+            {
+                return false;
+            }
+
+            // Check if any dynamic and static file dependency is sourceChangeAffectedOutputFiles
+            var hasAffected = sourceChangeAffectedOutputFiles.Intersect(dynamicallyObservedFiles).Any()
+                || sourceChangeAffectedOutputFiles.Intersect(Dependencies.Select(d => d.Path)).Any();
+
+            if (hasAffected)
+            {
+                return true;
+            }
+
+            if (dynamicallyObservedEnumerations.Any() || DirectoryDependencies.Any())
+            {
+                // Check if dynamic and static directory denpendencis under any directory of sourceChangeAffectedOutputDirectroies

```

Comment:
typo

Reply:
I noticed a few similar typo's, you might want to do a full text search. 

---
In reply to: [314827920](https://github.com/microsoft/BuildXL/pull/748#discussion_r314827920) [](ancestors = 314827920)

Reply:
Thanks